### PR TITLE
Intel GPU oneDNN upstreaming for library compilation

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
# Motivation
As proposed  in https://github.com/pytorch/pytorch/issues/114848 and https://github.com/pytorch/pytorch/issues/114723, oneDNN library is an important component for Intel GPU software ecosystem. 

This PR is intended to enable oneDNN compilation for Intel GPU.  It is the first step for we enabling any operators like `at::baddmm`.
With this PR, a static library `libdnnl.a` for GPU would be compiled in directory `/build/xpumkldnn_proj-prefix`.  It can be further linked to `libtorch_xpu.so` in future. The compilation would  depend on `USE_XPU` bool variables and runtime check like SYCL, which is defined in https://github.com/pytorch/pytorch/pull/116019 for runtime support. Once the #116019 merged, the compilation should be able to be triggered.

The modification is independent to oneDNN CPU compilation, hence no modification would be introduced for CPU Cmakefiles(e.g. FindMKLDNN.cmake)

@EikanWang @gujinghui @jgong5 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @LucasLLC

### CODE REVIEW

**Review by EikanWang — CHANGES_REQUESTED** [review]
In general, let's refine the code align with CPU build and check if we can merge the CMake to CPU.

File: cmake/External/XPUMKLDNN.cmake [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    -G ${CMAKE_GENERATOR} # Align Generator to Torch
+    BUILD_COMMAND ${MAKE_COMMAND}
+    INSTALL_COMMAND ""
+)
+
+ExternalProject_Get_Property(xpumkldnn_proj BINARY_DIR)
+set(__XPU__MKLDNN_BUILD_DIR ${BINARY_DIR})
+set(XPU_MKLDNN_LIBRARIES ${__XPU_MKLDNN_BUILD_DIR}/src/libdnnl.a)
+set(XPU_MKLDNN_INCLUDE ${PROJECT_SOURCE_DIR}/third_party/ideep/mkl-dnn/include)
+set(xpumkldnn_dep xpumkldnn_proj)
```
[the remark this answers is not in the document]
Comment (jgong5) [reply cmake/External/XPUMKLDNN.cmake:30]: yes, it makes sense to add underscore after "xpu" for all the symbol names.

File: cmake/External/XPUMKLDNN.cmake [hunk 2]
```diff
@@ -0,0 +1,36 @@
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    set(DNNL_DPCPP_HOST_COMPILER "g++")
+    # g++ is soft linked to /usr/bin/cxx, oneDNN woule not treat it as an absolute path
```
[the remark this answers is not in the document]
Comment (jgong5) [reply cmake/External/XPUMKLDNN.cmake:3]: Why do we have to treat GNU separately? Can you explain?

[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply cmake/External/XPUMKLDNN.cmake:3]: Sorry for the confusion here. 
The issue here is build option DNNL_HOST_COMPILER cannot be properly handle when it is `c++` (soft link `g++` created by system). oneDNN would complain that, `cxx` is not an absolute path "The valid values for DNNL_DPCPP_HOST_COMPILER: DEFAULT and g++ or absolute path to it"

**Comment by kit1980** [comment]
Too many people are notified on this PR, please create a new one.

**Pull request closed without merging**